### PR TITLE
environmentd: prepare to integrate 0dt upgrades with deploy generations

### PR DESCRIPTION
**Pull request by benesch**
Various refactorings that will allow the cloud orchestration layer to perform a 0dt upgrade if a feature flag is enabled for the environment.

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

Go commit by commit.

If it's helpful, here's the commit that I'm working on that stacks on top of this refactoring to actually perform the 0dt preflight checks: https://github.com/MaterializeInc/materialize/pull/27819/commits/99c9367d899bc36ae6f2604a281a494a6642d288

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

File: src/environmentd/src/deployment.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+//!
+//!   5. The new deployment completes any remaining initialization tasks and
+//!      prepares to serve queries.
+//!
+//!   6. Simultaneously, the external orchestrator tears down the old
+//!      deployment.
+//!
+//! [^1] "Configuration" here means "command-line flags". Most settings in an
+//! environment are changeable at runtime via LaunchDarkly feature flags, but
+//! occasionally
```
Comment (aljoscha) [inline src/environmentd/src/deployment.rs:44]: the suspense!

File: src/environmentd/src/deployment/state.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+/// will determine when promotion occurs. The future returned by
+/// `set_ready_to_promote` will resolve when promotion has occurred and the
+/// deployment should take over from the prior generation and begin serving
+/// queries.
+pub struct DeploymentState {
+    inner: Arc<Mutex<DeploymentStateInner>>,
+}
+
+impl DeploymentState {
+    /// Creates a new `LeaderState` for a deployment. The
```
Comment (aljoscha) [inline src/environmentd/src/deployment/state.rs:50]: trailing "The"

**Review by aljoscha — APPROVED** [review]
The newly factored deployment state handling is 🔥!

I did flag two trailing comments. And I did check out https://github.com/MaterializeInc/materialize/pull/27819/commits/99c9367d899bc36ae6f2604a281a494a6642d288 and thought about how it would play with bringing up an env in read-only mode, and things _should_ work out.

**Comment by benesch** [comment]
> I did flag two trailing comments.

Ugh! Thanks. I'm going to merge as is because the build is green, and then fix in #27819.

**Pull request merged**